File: 3043-find-the-length-of-the-longest-common-prefix/solution.cpp
#include <vector>
#include <unordered_set>
#include <algorithm>
#include <string>

class Solution {
public:
    int longestCommonPrefix(std::vector<int>& arr1, std::vector<int>& arr2) {
        std::unordered_set<int> prefixes;
        
        // Optimizing hash set capacity to avoid rehashing overhead
        prefixes.reserve(arr1.size() * 8);
        
        // Step 1: Insert all prefixes of arr1 elements into the hash set
        for (int val : arr1) {
            while (val > 0) {
                prefixes.insert(val);
                val /= 10;
            }
        }
        
        int max_len = 0;
        
        // Step 2: For each element in arr2, find its longest matching prefix
        for (int val : arr2) {
            while (val > 0) {
                if (prefixes.count(val)) {
                    // Since we go from longest to shortest, the first match is the longest
                    max_len = std::max(max_len, (int)std::to_string(val).length());
                    break;
                }
                val /= 10;
            }
        }
        
        return max_len;
    }
};

File: 3043-find-the-length-of-the-longest-common-prefix/solution.py
class Solution:
    def longestCommonPrefix(self, arr1: list[int], arr2: list[int]) -> int:
        prefixes = set()
        
        # Step 1: Add all possible prefixes from arr1 into the set
        for val in arr1:
            while val > 0:
                prefixes.add(val)
                val //= 10
                
        max_len = 0
        
        # Step 2: Check prefixes from arr2, going from longest to shortest
        for val in arr2:
            length = len(str(val))
            
            # Pruning: skip if the number of digits is already smaller than or equal to max_len
            if length <= max_len:
                continue
                
            while val > 0:
                if val in prefixes:
                    max_len = max(max_len, length)
                    break # First match going downwards is always the longest possible for this number
                
                val //= 10
                length -= 1
                
                # Early exit: no need to search shorter prefixes if they cannot beat max_len
                if length <= max_len:
                    break
                    
        return max_len